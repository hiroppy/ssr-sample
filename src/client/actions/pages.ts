import { State } from '../reducers/pages';

export type Actions =
  | ReturnType<typeof setEnv>
  | ReturnType<typeof setBaseUrl>
  | ReturnType<typeof resetPageStatus>
  | ReturnType<typeof loadAppProcess>
  | ReturnType<typeof loadAppProcessSuccess>
  | ReturnType<typeof loadAppProcessFailure>
  | ReturnType<typeof loadTopPage>
  | ReturnType<typeof loadTopPageSuccess>
  | ReturnType<typeof loadTopPageFailure>
  | ReturnType<typeof loadSagaPage>
  | ReturnType<typeof loadSagaPageSuccess>
  | ReturnType<typeof loadSagaPageFailure>
  | ReturnType<typeof loadApolloPage>
  | ReturnType<typeof loadApolloPageSuccess>
  | ReturnType<typeof loadApolloPageFailure>;

export const SET_ENV = 'SET_ENV';
export const setEnv = (env: State['env']) =>
  ({
    type: SET_ENV,
    payload: {
      env
    }
  } as const);

export const SET_BASE_URL = 'SET_BASE_URL';
export const setBaseUrl = (baseUrl: State['baseUrl']) =>
  ({
    type: SET_BASE_URL,
    payload: {
      baseUrl
    }
  } as const);

export const RESET_PAGES_STATUS = 'RESET_PAGES_STATUS';
export const resetPageStatus = () =>
  ({
    type: RESET_PAGES_STATUS
  } as const);

export const LOAD_APP_PROCESS = 'LOAD_APP_PROCESS';
export const loadAppProcess = () =>
  ({
    type: LOAD_APP_PROCESS
  } as const);

export const LOAD_APP_PROCESS_SUCCESS = 'LOAD_APP_PROCESS_SUCCESS';
export const loadAppProcessSuccess = () =>
  ({
    type: LOAD_APP_PROCESS_SUCCESS
  } as const);

export const LOAD_APP_PROCESS_FAILURE = 'LOAD_APP_PROCESS_FAILURE';
export const loadAppProcessFailure = (err: Error) =>
  ({
    type: LOAD_APP_PROCESS_FAILURE,
    payload: {
      err
    }
  } as const);

export const LOAD_TOP_PAGE = 'LOAD_TOP_PAGE';
export const loadTopPage = () =>
  ({
    type: LOAD_TOP_PAGE
  } as const);

export const LOAD_TOP_PAGE_SUCCESS = 'LOAD_TOP_PAGE_SUCCESS';
export const loadTopPageSuccess = () =>
  ({
    type: LOAD_TOP_PAGE_SUCCESS
  } as const);

export const LOAD_TOP_PAGE_FAILURE = 'LOAD_TOP_PAGE_FAILURE';
export const loadTopPageFailure = (err: Error) =>
  ({
    type: LOAD_TOP_PAGE_FAILURE,
    payload: {
      err
    }
  } as const);

export const LOAD_SAGA_PAGE = 'LOAD_SAGA_PAGE';
export const loadSagaPage = (maxLength: string | null) =>
  ({
    type: LOAD_SAGA_PAGE,
    payload: {
      maxLength
    }
  } as const);

export const LOAD_SAGA_PAGE_SUCCESS = 'LOAD_SAGA_PAGE_SUCCESS';
export const loadSagaPageSuccess = () =>
  ({
    type: LOAD_SAGA_PAGE_SUCCESS
  } as const);

export const LOAD_SAGA_PAGE_FAILURE = 'LOAD_SAGA_PAGE_FAILURE';
export const loadSagaPageFailure = (err: Error) =>
  ({
    type: LOAD_SAGA_PAGE_FAILURE,
    payload: {
      err
    }
  } as const);

export const LOAD_APOLLO_PAGE = 'LOAD_APOLLO_PAGE';
export const loadApolloPage = () =>
  ({
    type: LOAD_APOLLO_PAGE
  } as const);

export const LOAD_APOLLO_PAGE_SUCCESS = 'LOAD_APOLLO_PAGE_SUCCESS';
export const loadApolloPageSuccess = () =>
  ({
    type: LOAD_APOLLO_PAGE_SUCCESS
  } as const);

export const LOAD_APOLLO_PAGE_FAILURE = 'LOAD_APOLLO_PAGE_FAILURE';
export const loadApolloPageFailure = (err: Error) =>
  ({
    type: LOAD_APOLLO_PAGE_FAILURE,
    payload: {
      err
    }
  } as const);
