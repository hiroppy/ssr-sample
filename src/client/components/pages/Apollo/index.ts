import { Apollo } from './Apollo';

export default Apollo;
