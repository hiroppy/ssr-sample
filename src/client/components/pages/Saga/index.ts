import { Saga } from './Saga';

export default Saga;
