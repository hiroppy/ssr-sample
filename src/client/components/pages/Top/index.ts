import { Top } from './Top';

export default Top;
