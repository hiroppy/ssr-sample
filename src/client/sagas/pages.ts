import { END } from 'redux-saga';
import { put, call, take, select, takeLatest } from 'redux-saga/effects';
import {
  setEnv,
  resetPageStatus,
  loadAppProcess as LoadAppProcess,
  loadAppProcessSuccess,
  loadAppProcessFailure,
  loadTopPage as LoadTopPage,
  loadTopPageSuccess,
  loadTopPageFailure,
  loadSagaPage as LoadSagaPage,
  loadSagaPageSuccess,
  loadSagaPageFailure,
  loadApolloPage as LoadApolloPage,
  loadApolloPageSuccess,
  loadApolloPageFailure,
  LOAD_APP_PROCESS,
  LOAD_TOP_PAGE,
  LOAD_SAGA_PAGE,
  LOAD_APOLLO_PAGE
} from '../actions/pages';
import {
  fetchSagaCode,
  fetchSagaCodeSuccess,
  fetchSagaCodeFailure,
  FETCH_SAGA_CODE_SUCCESS,
  FETCH_SAGA_CODE_FAILURE
} from '../actions/fetchSaga';

// e.g. write common processing to be performed on all pages
// don't call `stopSaga`
function* appProcess(actions: ReturnType<typeof LoadAppProcess>) {
  yield put(setEnv(process.env.NODE_ENV || 'development'));
  yield put(loadAppProcessSuccess());
}

// if you run async process, you have to change the code like below
// function* appProcess(actions: ReturnType<typeof LoadAppProcess>) {
//   try {
//     // async

//     yield put(loadAppProcessSuccess());
//   } catch (err) {
//     yield put(loadAppProcessFailure(err));
//   } finally {
//     if (!process.env.IS_BROWSER) {
//       yield call(stopSaga);
//     }
//   }
// }

function* loadTopPage(actions: ReturnType<typeof LoadTopPage>) {
  yield changePage();
  yield put(loadTopPageSuccess());

  if (!process.env.IS_BROWSER) {
    yield call(stopSaga);
  }
}

function* loadSagaPage(actions: ReturnType<typeof LoadSagaPage>) {
  try {
    yield changePage();
    yield put(fetchSagaCode(actions.payload.maxLength));

    const res:
      | ReturnType<typeof fetchSagaCodeSuccess>
      | ReturnType<typeof fetchSagaCodeFailure> = yield take([
      FETCH_SAGA_CODE_SUCCESS,
      FETCH_SAGA_CODE_FAILURE
    ]);

    if (res.type === FETCH_SAGA_CODE_FAILURE) {
      throw res.payload.err;
    }

    yield put(loadSagaPageSuccess());
  } catch (err) {
    yield put(loadSagaPageFailure(err));
  } finally {
    if (!process.env.IS_BROWSER) {
      yield call(stopSaga);
    }
  }
}

function* loadApolloPage(actions: ReturnType<typeof LoadApolloPage>) {
  yield changePage();
  yield put(loadApolloPageSuccess());

  if (!process.env.IS_BROWSER) {
    yield call(stopSaga);
  }
}

function* stopSaga() {
  yield put(END);
}

function* changePage() {
  // don't need to call resetPageStatus because baseUrl is required by fetch on Node.js environment,
  // also state has already been initialized at this time
  if (process.env.IS_BROWSER) {
    yield put(resetPageStatus());
  }
}

export function* pagesProcess() {
  yield takeLatest(LOAD_APP_PROCESS, appProcess);
  yield takeLatest(LOAD_TOP_PAGE, loadTopPage);
  yield takeLatest(LOAD_SAGA_PAGE, loadSagaPage);
  yield takeLatest(LOAD_APOLLO_PAGE, loadApolloPage);
}
