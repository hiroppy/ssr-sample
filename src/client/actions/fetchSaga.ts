import { loadSagaPage } from './pages';
import { sagaSamples } from '../../server/responseSchema';

export const FETCH_SAGA_CODE = 'FETCH_SAGA_CODE';
export const fetchSagaCode = (maxLength: Parameters<typeof loadSagaPage>[0]) =>
  ({
    type: FETCH_SAGA_CODE,
    payload: {
      maxLength
    }
  } as const);

export const FETCH_SAGA_CODE_SUCCESS = 'FETCH_SAGA_CODE_SUCCESS';
export const fetchSagaCodeSuccess = (samples: typeof sagaSamples) =>
  ({
    type: FETCH_SAGA_CODE_SUCCESS,
    payload: {
      samples
    }
  } as const);

export const FETCH_SAGA_CODE_FAILURE = 'FETCH_SAGA_CODE_FAILURE';
export const fetchSagaCodeFailure = (err: Error) =>
  ({
    type: FETCH_SAGA_CODE_FAILURE,
    payload: {
      err
    }
  } as const);

export const ADD_LIKE = 'ADD_LIKE';
export const addLike = (id: number) =>
  ({
    type: ADD_LIKE,
    payload: {
      id
    }
  } as const);

export const ADD_LIKE_SUCCESS = 'ADD_LIKE_SUCCESS';
export const addLikeSuccess = (res: typeof sagaSamples[0]) =>
  ({
    type: ADD_LIKE_SUCCESS,
    payload: {
      res
    }
  } as const);

export const ADD_LIKE_FAILURE = 'ADD_LIKE_FAILURE';
export const addLikeFailure = (err: Error) =>
  ({
    type: ADD_LIKE_FAILURE,
    payload: {
      err
    }
  } as const);

export type Actions =
  | ReturnType<typeof fetchSagaCode>
  | ReturnType<typeof fetchSagaCodeSuccess>
  | ReturnType<typeof fetchSagaCodeFailure>
  | ReturnType<typeof addLike>
  | ReturnType<typeof addLikeSuccess>
  | ReturnType<typeof addLikeFailure>;
