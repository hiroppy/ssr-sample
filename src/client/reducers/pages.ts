import { Actions } from '../actions/pages';

export type State = {
  isLoadingCompletion: boolean;
  error: Error | null;
  baseUrl: string;
  env: typeof process.env.NODE_ENV;
};

export const initialState: State = {
  isLoadingCompletion: false,
  error: null,
  baseUrl: '',
  env: 'development'
};

export const reducer = (state: State = initialState, action: Actions): State => {
  switch (action.type) {
    case 'SET_ENV':
      return { ...state, env: action.payload.env };
    case 'RESET_PAGES_STATUS':
      return initialState;
    case 'SET_BASE_URL':
      return { ...state, baseUrl: action.payload.baseUrl };
    case 'LOAD_TOP_PAGE':
    case 'LOAD_SAGA_PAGE':
    case 'LOAD_APOLLO_PAGE':
      return { ...state, isLoadingCompletion: false };
    case 'LOAD_TOP_PAGE_SUCCESS':
    case 'LOAD_SAGA_PAGE_SUCCESS':
    case 'LOAD_APOLLO_PAGE_SUCCESS':
      return { ...state, isLoadingCompletion: true };
    case 'LOAD_TOP_PAGE_FAILURE':
    case 'LOAD_SAGA_PAGE_FAILURE':
    case 'LOAD_APOLLO_PAGE_FAILURE':
      return { ...state, isLoadingCompletion: true, error: action.payload.err };
    default:
      return state;
  }
};
