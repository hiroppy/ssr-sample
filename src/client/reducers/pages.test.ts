import { reducer, initialState } from './pages';
import * as actions from '../actions/pages';

test('should return the initial state', () => {
  expect(reducer(undefined, { type: undefined } as any)).toEqual(initialState);
});

test('should handle RESET_PAGES_STATUS', () => {
  expect(
    reducer({ ...initialState, baseUrl: 'test' }, actions.resetPageStatus())
  ).toMatchSnapshot();
});

test('should handle SET_ENV', () => {
  expect(reducer(initialState, actions.setEnv('test'))).toMatchSnapshot();
});

test('should handle SET_BASE_URL', () => {
  expect(reducer(initialState, actions.setBaseUrl('base-url'))).toMatchSnapshot();
});

test('should handle LOAD_TOP_PAGE', () => {
  expect(reducer(initialState, actions.loadTopPage())).toMatchSnapshot();
});

test('should handle LOAD_TOP_PAGE_SUCCESS', () => {
  expect(reducer(initialState, actions.loadTopPageSuccess())).toMatchSnapshot();
});

test('should handle LOAD_TOP_PAGE_FAILURE', () => {
  expect(reducer(initialState, actions.loadTopPageFailure(new Error('404')))).toMatchSnapshot();
});

test('should handle LOAD_SAGA_PAGE', () => {
  expect(reducer(initialState, actions.loadSagaPage('len'))).toMatchSnapshot();
});

test('should handle LOAD_SAGA_PAGE_SUCCESS', () => {
  expect(reducer(initialState, actions.loadSagaPageSuccess())).toMatchSnapshot();
});

test('should handle LOAD_SAGA_PAGE_FAILURE', () => {
  expect(reducer(initialState, actions.loadSagaPageFailure(new Error('404')))).toMatchSnapshot();
});

test('should handle LOAD_APOLLO_PAGE', () => {
  expect(reducer(initialState, actions.loadApolloPage())).toMatchSnapshot();
});

test('should handle LOAD_APOLLO_PAGE_SUCCESS', () => {
  expect(reducer(initialState, actions.loadApolloPageSuccess())).toMatchSnapshot();
});

test('should handle LOAD_APOLLO_PAGE_FAILURE', () => {
  expect(reducer(initialState, actions.loadApolloPageFailure(new Error('404')))).toMatchSnapshot();
});
