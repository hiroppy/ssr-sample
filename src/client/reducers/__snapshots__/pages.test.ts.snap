// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle LOAD_APOLLO_PAGE 1`] = `
Object {
  "baseUrl": "",
  "env": "development",
  "error": null,
  "isLoadingCompletion": false,
}
`;

exports[`should handle LOAD_APOLLO_PAGE_FAILURE 1`] = `
Object {
  "baseUrl": "",
  "env": "development",
  "error": [Error: 404],
  "isLoadingCompletion": true,
}
`;

exports[`should handle LOAD_APOLLO_PAGE_SUCCESS 1`] = `
Object {
  "baseUrl": "",
  "env": "development",
  "error": null,
  "isLoadingCompletion": true,
}
`;

exports[`should handle LOAD_SAGA_PAGE 1`] = `
Object {
  "baseUrl": "",
  "env": "development",
  "error": null,
  "isLoadingCompletion": false,
}
`;

exports[`should handle LOAD_SAGA_PAGE_FAILURE 1`] = `
Object {
  "baseUrl": "",
  "env": "development",
  "error": [Error: 404],
  "isLoadingCompletion": true,
}
`;

exports[`should handle LOAD_SAGA_PAGE_SUCCESS 1`] = `
Object {
  "baseUrl": "",
  "env": "development",
  "error": null,
  "isLoadingCompletion": true,
}
`;

exports[`should handle LOAD_TOP_PAGE 1`] = `
Object {
  "baseUrl": "",
  "env": "development",
  "error": null,
  "isLoadingCompletion": false,
}
`;

exports[`should handle LOAD_TOP_PAGE_FAILURE 1`] = `
Object {
  "baseUrl": "",
  "env": "development",
  "error": [Error: 404],
  "isLoadingCompletion": true,
}
`;

exports[`should handle LOAD_TOP_PAGE_SUCCESS 1`] = `
Object {
  "baseUrl": "",
  "env": "development",
  "error": null,
  "isLoadingCompletion": true,
}
`;

exports[`should handle RESET_PAGES_STATUS 1`] = `
Object {
  "baseUrl": "",
  "env": "development",
  "error": null,
  "isLoadingCompletion": false,
}
`;

exports[`should handle SET_BASE_URL 1`] = `
Object {
  "baseUrl": "base-url",
  "env": "development",
  "error": null,
  "isLoadingCompletion": false,
}
`;

exports[`should handle SET_ENV 1`] = `
Object {
  "baseUrl": "",
  "env": "test",
  "error": null,
  "isLoadingCompletion": false,
}
`;
